cpp/src/arrow/table.cc

```diff
@@ -150,6 +154,21 @@ Status Column::ValidateData() {
   return Status::OK();
 }
 
+Status Column::cast(const std::shared_ptr<DataType>& type, std::shared_ptr<Column>* out) {
+  std::vector<std::shared_ptr<Array>> new_data;
+  std::shared_ptr<ChunkedArray> chunks = this->data();
+  for (int ix = 0; ix < chunks->num_chunks(); ix++) {
+    std::shared_ptr<Array> chunk = chunks->chunk(ix);
+    std::shared_ptr<Array> converted_chunk;
+    RETURN_NOT_OK(chunk->cast(type, &converted_chunk));
```

Comment:
Per above, I'm not quite comfortable adding these instance methods yet. How about adding an `EncodeDictionary` function for `Column` also? 

This is also problematic because different chunks might have different dictionaries -- this further supports handling casting to dictionary as a special case, and we can contemplate more general casting later. I would need to look more closely at the DictionaryBuilder, but it will need to retain a common dictionary so that all chunks have the same dictionary. You will have to create a vector of dictionary index chunks and then wrap the indices into a DictionaryArray at the end once you have the dictionary for the whole column